5010 redesign product form

isaacfreeman opened the pull request.

BenMorganIO committed:
Bootstrapping Email HTML templates.

jhawthorn committed:
Add Spree instrumentation

Add instrumentation to some checkout methods

Fixes #4802

brchristian committed:
Use slower but robust ActiveRecord for migration

When I ran this migration on my 2-1-stable store, I got the following exception:

    StandardError: An error has occurred, all later migrations canceled:

    Mysql2::Error: Column 'shipment_total' cannot be null: UPDATE spree_orders
                 SET shipment_total = (SELECT SUM(spree_shipments.cost) AS sum_id
                                       FROM spree_shipments
                                       WHERE spree_shipments.order_id = spree_orders.id)
                 WHERE spree_orders.completed_at IS NOT NULL AND spree_orders.shipment_total = 0/Users/username/appname/db/migrate/20140609012476_set_shipment_total_for_users_upgrading.spree.rb:7:in `up'
    ActiveRecord::StatementInvalid: Mysql2::Error: Column 'shipment_total' cannot be null: UPDATE spree_orders
                 SET shipment_total = (SELECT SUM(spree_shipments.cost) AS sum_id
                                       FROM spree_shipments
                                       WHERE spree_shipments.order_id = spree_orders.id)
                 WHERE spree_orders.completed_at IS NOT NULL AND spree_orders.shipment_total = 0

Rephrasing the raw SQL into ActiveRecord, the migration ran slower but didn’t crash.

(I was intending to add `if shipments && shipments.sum(:cost)` to line 7 but surprisingly didn’t need to.)

Fixes #4823

Commit:
Fix email specs.

dunric committed:
Spree::Zone#country_list fix

Just fixing wrong default value of case expression - should return an Array instance, ie. `[]` instead of `nil`.

Fixes #4827

brchristian committed:
Add state_text to StockLocation

I'm working on a revision to `Spree::TaxCloud`, and we use the `StockLocation` addresses pretty heavily. It'd be a huge help to have the `state_text` method that exists in the [`Spree::Address` model](https://github.com/spree/spree/blob/master/core/app/models/spree/address.rb#L39).

Fixes #4872

Senjai committed:
Encapsulate handling of payment preconditions

This would allow us to create methods that would verify the integrity of
a payment but allow us to do custom things with it.

Add utility methods for explicit authorization and capture.

There are times in custom applications that we explicitly want to
authorize or capture regardless of the payment_methods auto capture
preference or the store configuration. This lets us do that while
maintaining that the regular preconditions for payments are met.

Refactor process_payments!

As with the payment changes, this allows us to call custom methods on
payments for processing.

Add authorize_payments! and capture_payments! to orders.

Allows us to specifically authorize or capture payments on an order.

Move payment processing to a concern

Fixes #4879

BenMorganIO committed:
remove top and bottom pixels from admin nav
fixes #4924

jordan-brough committed:
Make best-promotion query more deterministic

Mostly for the benefit of the specs under MySQL.

This query was generating some sporadic spec failures for me when using MySQL as the database. This is because MySQL does not store fractional seconds so the "created_at" was sometimes the same value for two adjustments when running specs. And when created_at is identical the query below would actually pick the earlier adjustment instead of the more recent one.  Adding "id" as a fallback ordering fixes this problem and makes the behavior more deterministic.

The issue is reproduceable in postgres as well by using Timecop.

Fixes #4943

BenMorganIO committed:
Add Canadian provinces. Fixes #5016

Commit:
Add test rake task and refactor layout.

Commit:
Cleanup unused premailer/ink stuff.

Commit:
Cleanup templates.

Commit:
Reorganised product form

Commit:
Inline vriant editing

- variant edit form appears inline in the variant table
- variant action URLs fixed

Commit:
Inline vriant editing

- variant edit form appears inline in the variant table
- variant action URLs fixed

Commit:
Moved variant table rows to their own partial

Commit:
Editing variants for product

- product accepts_nested-attributes_for variants
- merged variant partial and variant form partial (not happy with this as a final result)
- remove variants list from HTML unless suer can admin variants
- simplified variants/edit.js.erb to just hide/show the form for each variant
- reverted variants/edit.html.erb, so everything works again if you go to the old route for editing variants

Commit:
Editing variants for product

- product accepts_nested-attributes_for variants
- merged variant partial and variant form partial (not happy with this as a final result)
- remove variants list from HTML unless suer can admin variants
- simplified variants/edit.js.erb to just hide/show the form for each variant
- reverted variants/edit.html.erb, so everything works again if you go to the old route for editing variants

Commit:
Use Rails environments instead of DB configurations

In the admin forms for Spree::PaymentMethod and Spree::Tracker the collection used for the environment field is the database configuration keys. This is wrong and it becomes quite apparent in the case of applications that connect to more than one database. This commit replaces said collection with a collection of the available environment configurations in '/config/environments/'.

Closes #5170

Commit:
Simple fix for payment source.

Steps to reproduce the issue:

1. Add an order through the admin interface
2. Attempt to pay via cheque for a user who has a credit card number
stored in their profile.
3. Submit the form

This will 500, because params[:card] is set from the form on the initial
page for the radio button from the credit card profiles.

This is a simple, though incomplete fix for the issue. A proper fix
would be to change the form so that params[:card] is qualified by the
payment_method which it belongs.

There are also likely issues with multiple payment methods which support
profiles, as there is likely to be confusion between them. This should
suffice for now.

Fixes #5209

Commit:
Go to cart and then call next, to trigger hooks
Fixes #5216

jordan-brough committed:
Add API endpoint /api/shipments/mine

With exchanges and split shipments we now have times when stores will be interested in querying a user's shipments directly rather than having to go through the user's orders.

Note: Trying to prevent a bunch of n+1 queries with a bunch of includes.

Fixes #5219

HoyaBoya committed:
promotions api show

Closes #5233

HoyaBoya committed:
provide status code on promotion handler

This change adds a status_code to the JSON response for the promotion code response object. It's a little bit brittle to bind on the translation message from ```Spree.t``` and sometimes more granularity is needed then success/error. The status_code should hopefully solve this.

Closes #5234

Commit:
Improved variant editing

- add variants
- variants can be opened for editing
- when variant fields are updated, also update corresponding elements so the fields will be correct when the edit section is closed again
- modified link_to_add_field helper method to handle elements other than tr (in this case, we're using tbody as a wrapper for each variant)
-

Commit:
Improved variant editing

- add variants
- variants can be opened for editing
- when variant fields are updated, also update corresponding elements so the fields will be correct when the edit section is closed again
- modified link_to_add_field helper method to handle elements other than tr (in this case, we're using tbody as a wrapper for each variant)
-

Commit:
Manage properties from product form

- product form fieldset for properties (invisible until toggled by properties button)
- _propduct_properties partial (largely duplicates product_properties/index.html, but I'm keeping the latter around for now)
- set up properties in product controller
- added id passthrough to link_to_add_fields helper method, and specify id in select.js.erb, so we don't add fields to other parts of the form
- adjusted link_to_delete in _product_property_fields partial to specify URL
- removed unused _properties_form partial
- moved variants code from product form to partials _variants and master_variant_fields

Commit:
Adjusted layout for product form

- more consistent HTML for inline toolbars
- tweaked column widths for variants
- moved Manage Variants button to product details section (not sure about this, may move it back)
- moved images section to a partial

Commit:
Adjusted layout for product form

- more consistent HTML for inline toolbars
- tweaked column widths for variants
- moved Manage Variants button to product details section (not sure about this, may move it back)
- moved images section to a partial

Commit:
Moved Stock and SEO fields to own partials

Commit:
Moved Stock and SEO fields to own partials

Commit:
JS improvements

- moved option_types to own partial
- moved variants table to own partial
- separate JS method for adding variants
- reorder variants
- show delete/active variants
- improved new variant JS

Commit:
JS improvements

- moved option_types to own partial
- moved variants table to own partial
- separate JS method for adding variants
- reorder variants
- show delete/active variants
- improved new variant JS

Commit:
JS improvements

- moved option_types to own partial
- moved variants table to own partial
- separate JS method for adding variants
- reorder variants
- show delete/active variants
- improved new variant JS

Commit:
Update variants table when option types change

- product_slug data in product edit view
- updateOptionTypes() and updateVaraintsTable() functions in admin.js.erb
- update method in products controller checks for XHR request (there's probably a better way to do this)
- index method in variants controller returns variants table
- cosmetic improvements to variants table

Commit:
Mark variants for deletion

- variants table style and layout improvements
- mark variants for deletion when product form is submitted instead of deleting immediately

Commit:
Automatically check the return item checkbox when interacting with settings for returning the item

Fixes #5272

Commit:
Merge branch '5010-redesign-product-form' of github.com:isaacfreeman/spree into 5010-redesign-product-form

Conflicts:
	backend/app/assets/javascripts/spree/backend/admin.js.erb
	backend/app/assets/stylesheets/spree/backend/sections/_products.scss

Commit:
Fixing support for remembering your cart after logging in as well as remembering your last order you were working on after logging out and then back in again.

Fixes #5075
Fixes #5273
Fixes spree/spree_auth_devise#220

metade committed:
Refactor Spree::OrderContents#add_to_line_item so that the logic of setting the line item price based on the product options is reusable when creating an updating a line item

Closes #5274

Making Spree::LineItem#options= more robust for vanilla line items.

Using assign_attributes rather than iterating through the hash and setting the option values manually

Commit:
Fix using check payment method in admin.

metade committed:
Extracting Spree::ShippingMethod.find_by_name from create_shipments_from_params because it should be Spree::ShippingMethod.find_by_admin_name but this change needs to be handled carefully

Closes #5275

valpackett committed:
add meta title to products

Closes #5276

valpackett committed:
make 'Manage Variants' translatable

Closes #5277

Commit:
Add missing translation.

metade committed:
Removing unnecessary controllers and views from the admin
Fixes #5280

Commit:
Add remibursement credits to admin panel under refunds

Closes #5283

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
+  $(correspondingShowElement).html( editField.children('option:selected').text() );
+}
+
+function replaceId( element, newId ) {
```
Extra spaces around the parameters here. Other methods do not have this style.

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
+}
+
+function replaceId( element, newId ) {
+  element.prop("id", element.prop("id").replace(/\d+/, newId))
```
double quotes used here when singles quotes is mostly used elsewhere

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
+  element.prop("id", element.prop("id").replace(/\d+/, newId))
+}
+function updateOptionTypes() {
+  var product_slug = $('form.edit_product').data('product-slug');
```
underscored variable names used here when camel case is used elsewhere.

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -123,7 +123,79 @@ handle_date_picker_fields = function(){
+            url: "/admin/products/" + product_slug,
+            data: { product: { option_type_ids: option_type_ids }},
+            type: 'patch',
+            success: function( data ) {
```
Spaces around parameters. I won't comment on this any further on this as there are a number of examples. Pick a style and stick with it. Same with the camel case vs underscored.

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -154,11 +228,47 @@ $(document).ready(function(){
-    new_table_row.find("a").each(function () {
+    newElement.find("a").each(function () {
       var el = $(this);
       el.prop('href', '#');
```
this could just be:

```
$(this).prop('href', '#');
```

isaacfreeman replied:
Yeah, I changed that when I made a special version of this specific to variants, but I guess I've mucked around with this generic version enough to say I own it too.

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -154,11 +228,47 @@ $(document).ready(function(){
+    // When cloning a new row, set the href of all icons to be an empty "#"
+    // This is so that clicking on them does not perform the actions for the
+    // duplicated row
+    newVariant.find("a").each(function () {
```
Looks like a bit of duplication here with a several lines above. Anyway to put this stuff related to cloning in a reusable method?

isaacfreeman replied:
Yeah, I'll have another look at that. It's also violating my principle of keeping jQuery handler functions as short as possible. It got this way because I went back and forth on whether to write a new function for variants or try to modify the existing one. Will refactor both.

djones commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -271,6 +381,33 @@ $(document).ready(function(){
+          $("table.sortable tr:even").removeClass("odd even").addClass("even");
+          $("table.sortable tr:odd").removeClass("odd even").addClass("odd");
+        }
+
```
extra line here

djones commented on backend/app/assets/stylesheets/spree/backend/sections/_products.scss:
```diff
@@ -1,3 +1,10 @@
+form.edit_product {
```
You're not making use of the nesting here so you could just make this:

```
form.edit_product .inline-menu {
  float: right;
  margin-bottom: 10px;
}
```

djones commented on backend/app/assets/stylesheets/spree/backend/sections/_products.scss:
```diff
@@ -35,6 +60,97 @@
+        &:nth-child(6) {
+          border-right: 1px solid $color-border;
+        }
+
```
extra line here

djones commented on backend/app/assets/stylesheets/spree/backend/sections/_products.scss:
```diff
@@ -35,6 +60,97 @@
+
+      &.actions {
+        background-color: transparent;
+        border-right: none !important;
```
You could make this:

```
border: none !important;
border-left: 1px solid $color-2 !important;
```

Not sure if that's weird though.

isaacfreeman replied:
Not weird at all.

You can tell this is copy-and-pasted code because it's got !important in it. These styles should really be in the shared/_tables.scss, with more generic class names.

djones commented on backend/app/views/spree/admin/products/_form.html.erb:
```diff
@@ -177,3 +119,22 @@
 <% end %>
+
+
+<script>
```
this'd be better in a JS file somewhere...

isaacfreeman replied:
Another case of "should I fix this?" And again, I've changed this file so much now that there's no point in still being reluctant to move this.

djones commented on backend/app/views/spree/admin/products/_form.html.erb:
```diff
@@ -177,3 +119,22 @@
+  });
+  $( "a#add_variants" ).click(function( event ) {
+    event.preventDefault();
+    $( '#admin_product_form_multiple_variants' ).slideToggle();
```
I think you may be able to shorten this to just:

```
$( '#admin_product_form_multiple_variants, #admin_product_form_master_variant' ).slideToggle();
```

http://api.jquery.com/multiple-selector/

djones commented on backend/app/views/spree/admin/products/_product_properties.html.erb:
```diff
@@ -0,0 +1,41 @@
+  $('#product_properties').on('keydown', 'input.autocomplete', function() {
+    already_auto_completed = $(this).is('ac_input');
+    if (!already_auto_completed) {
+      $(this).autocomplete({source: properties});
```
might be able to chain this.

```
$(this).autocomplete({source: properties}).focus();
```

djones commented on backend/app/views/spree/admin/products/_product_properties.html.erb:
```diff
@@ -0,0 +1,41 @@
+  </tbody>
+</table>
+
+<%= javascript_tag do -%>
```
Would be good to put this in a JS file somewhere.

At very least this code doesn't wait until the dom is ready but it probably should do?

djones commented on backend/app/views/spree/admin/products/_seo_fields.html.erb:
```diff
@@ -0,0 +1,13 @@
+<div data-hook="admin_product_form_meta_keywords">
```
Duplication here. Make it so you can just pass in the field names `meta_keywords` and `meta_description` to have it render the same HTML.

isaacfreeman replied:
Disagree on this one. Having two separate groups with data-hooks makes it much easier to use Deface to modify the form. Having just had to deal with using Deface on views that didn't account for this in spree_static_content, this seems more important to me than being DRY.

djones commented on backend/app/views/spree/admin/prototypes/select.js.erb:
```diff
@@ -1,4 +1,5 @@
+<%# binding.remote_pry %>
```
remove this line

djones commented on backend/app/views/spree/admin/variants/_master_variant_fields.html.erb:
```diff
@@ -0,0 +1,60 @@
+</div>
+<div class="six columns">
+  <ul id="shipping_specs">
+    <li id="shipping_specs_weight_field" data-hook="admin_product_form_weight" class="field alpha omega six columns">
```
Some duplication over weight, height, width and depth. May not be worth putting into a partial though.

isaacfreeman replied:
Another one that's there for Deface, so it's easy to get at the particular fields.

I do find these dimension fields weird, as they're almost always left empty. I guess they're needed for some shipping methods, though, so they need to be present. It's a pain on the front end, because they take up most of the variant form with little value to most users. Still, that's another UX question for another day. I'm trying to keep focused on overall variant management for this PR.

djones commented on backend/app/views/spree/admin/variants/_variant.html.erb:
```diff
@@ -0,0 +1,106 @@
+        <td id="show_variant_<%= variant_counter %>_option_<%= option_type.name %>" class="option_value field-display"><%= variant.option_value(option_type.name) %></td>
+      <% end %>
+      <td>
+        <span id="show_product_variants_attributes_<%= variant_counter %>_height" class="field-display">
```
``` erb
<% ['height', 'width', 'depth'].each do |measurement| %>
  <span id="show_product_variants_attributes_<%= variant_counter %>_<%= measurement %>" class="field-display">
    <%= number_with_precision(variant.send(measurement), :precision => 2) %>
  </span><%= measurement.chars.first %>
  &times;
<% end %>
```

isaacfreeman replied:
Another case for Deface. 

Dimensions don't have any units, and I suspect that adding them in with Deface would be a common thing to do (if it isn't it should be). So really these need to be kept separate and have data-hooks added to them.

I do need to review this though, because it doesn't work on the front end – if there are no dimensions you get "h x w x d" with no numbers.

djones commented on core/app/models/spree/product.rb:
```diff
@@ -93,6 +93,8 @@ class Product < Spree::Base
 
     accepts_nested_attributes_for :product_properties, allow_destroy: true, reject_if: lambda { |pp| pp[:property_name].blank? }
 
+    accepts_nested_attributes_for :variants, allow_destroy: true, :reject_if => lambda { |v| v[:sku].blank? }
```
Ruby 1.9 syntax on `allow_destroy: true`. I think you've used mostly 1.8 elsewhere.

BenMorganIO committed:
call reimbursement with id method

Commit:
Tidying up

- more consistent code style in admin.js.erb
- moved product form jQuery handler code to admin.js.erb
- minor stylesheet changes
- removed commented pry call
- JS adjustment in product_proerties partial

peterberkenbosch committed:
add specs for rails_environments helper method introduced in #5170

HoyaBoya committed:
Allow for notification of API errors

Due to the way the rescue handlers are setup for the API, there's no easy way to alert Honeybadger/Airbrake/etc of an error in the API. Instead we have to check the client error JSON and debug from that which is not very ideal expecially if we don't control the client code. This is a simple change to let people decorate in a hook on ```error_during_processing``` so that we can notify on error.

This should help me get rid of one more ```class_eval``` and ```alias_method_chain``` in my application.

Thought this was more flexible then to have but open doing the KISS solution below:

```
if defined?(Honeybadger) then Honeybadger.notify()
```

incorporate feedback from jordan, make this proc based

Get rid of more class evals and let people use this as a config via Proc.

Closes #5285

peterberkenbosch committed:
implement the code from #4488 by @k-akarsh.

Closes #5286

Commit:
Don't allow creating a new reimbursement if there are any pending ones on the customer return

Closes #5287

camelmasa committed:
Refactor files

Update 1.8 hash style to 1.9 hash style.

Closes #5288

peterberkenbosch committed:
fix failing specs introduced with the min and max amounts on item total promotion rule.

peterberkenbosch committed:
we need to stub the environments file read before the specs run, since the memoization of the @@rails_environments is always returning []

camelmasa committed:
Fix not load environment files when test

Closes #5289

remove the stub since Rails.root is a waaaaay better way of testing this.

peterberkenbosch committed:
fix more failing specs due to the changes in the item total promotion rule (adding min and max preferences)

Commit:
Fix promo specs.

JDutil committed:
Merge pull request #4385 from freerunningtech/guides-tree-rewrite

Guides to be maintained in core repo now.

tomykaira committed:
Let PaymentMethod create a customer profile using a payment profile

In the original code, `payment_method.create_profile` is called only when the source has neither a customer profile nor a payment profile.
This patch removes that check in order to create a customer profile using a payment profile.

Closes #5292

camelmasa committed:
Create Spree::Core::ControllerHelpers::Auth module test

Closes #5293

camelmasa committed:
Create Spree::Core::ControllerHelpers::Search module test

Closes #5294

camelmasa committed:
Remove default_notification_payload method

It's unused method.

Closes #5295

BenMorganIO committed:
create stock location that passes validations

Closes #5296

camelmasa committed:
Create Spree::Core::ControllerHelpers::Order module's test
Fixes #5298

Commit:
Fix deprecation.

peterberkenbosch committed:
make sure we get the current order for use in the cart as well. Refactored some shared code as well.

Fixes #5300

camelmasa committed:
Create Spree::Core::ControllerHelpers::Store module test

camelmasa committed:
Remove unused file

Closes #5303

frankmt committed:
Removing non existent items from params to avoid bug when clicking on remove multiple times or when clicking on multiple remove buttons

Fixes #5304
Closes #5305

frankmt committed:
Removing non existent items from params to avoid bug when clicking on remove multiple times or when clicking on multiple remove buttons

Skipping filter if there is only one line item being updated

Fixing order controller spec

Dealing with case when line items attributes is nil

Fixes #5318

peterberkenbosch committed:
keep the simple_current_order simple :)

Paveltarno committed:
Dont create product if master variant invalid.
[Fixes #5306]
Fixes #5330

davidlibrera committed:
Fixes regressions in sprockets >= 2.1.4

Fixes #5311
Fixes #5313

bbuchalter committed:
Allow partial emails when searching for orders
Fixes #5314

Commit:
Fix spec.

Commit:
Revert "Removing non existent items from params to avoid bug when clicking on remove multiple times or when clicking on multiple remove buttons"

This reverts commit 466b3fecf71bc38714dd637e04a5bd4ac0a6cc5d.

This commit broke the build.

deodad committed:
Fix item_adjustment_spec.rb

Test after tax_adjustments callback instead of after promo_adjustments, which was testing a few lines earlier
Fixes #5320

Willianvdv committed:
Only cancel cancellable shipments.

Fixes #5321
Fixes #5299

camelmasa committed:
Update Ruby hash syntax to 1.9

Create Spree::Core::ControllerHelpers::StrongParameters test
Fixes #5322

Commit:
removes window.location.origin

isn't compatible with Internet Explorer

Fixes #5323

camelmasa committed:
:scissors:

Commit:
corrects fetch_cart urls
Fixes #5325

Commit:
Merge pull request #5324 from camelmasa/refactor-shipment_handler

Refactor Spree::ShipmentHandler

Commit:
Revert "Only cancel cancellable shipments."

This reverts commit d870c4e316e3aad6f63c728d224f6a16d4edde57.

Commit:
Fix form fields.

Commit:
Fix broken spec due to LineItem#options= not being able to assign target_shipment.

Commit:
Add self dependencies.

Fixes #5157
Fixes #5328

BenMorganIO committed:
added new lines at end of files

fixes issue #4690

remove whitespace

added branch option to main readme

update hierarchy in cmd readme

Closes #5331

Paveltarno committed:
addi18n to prom index
Fixes #5333

jordan-brough committed:
Have reimbursements touch their customer returns

For example, completing a reimbursement may mean that a customer return is now fully reimbursed, etc.
Fixes #5334

Commit:
Update jquery dep. refs #3203

ifan-godaddy committed:
Always call registered extension line item comparison hooks
Fixes #5343

Davidslv committed:
Update promotions.md

`Adjustment#determine_eligibility` && `Promotion#rules_are_eligible?` seems to have been deprecated at some point and no longer exist.

I tried to correct the text as best as I could.

Thank you

Fixes #5347

jordan-brough committed:
Detect old-style order refunds

In the `Spree::Order#has_non_reimbursement_related_refunds?` method.

The old style was not migrated due to not enough information being stored and because of potential ambiguities, but we still need to account for it here.
Fixes #5348

HoyaBoya committed:
Remove check for payment step and has payment step in checkout flow when updating payment source
Fixes #5351

Senjai committed:
Reduce queries whenever _products are rendered.
Fixes #5352

HoyaBoya committed:
require at least one valid payment on transition to complete
Fixes #5350

Commit:
Don't raise an error on credit card default setting of another card is expired
Fixes #5353

Commit:
RSpec 3.1 support & Fix a couple Rails 4.1.6 regressions.

Commit:
Remove typo

HoyaBoya committed:
let value have validation of db column length

need to rescue AR issue for spree install, ci, etc (chicken and the egg issue)

do validations as a re-useable validtor

explicit argument error message
Fixes #5361

alexandremcosta committed:
better way to redefine existing methods: allows to call original behaviour with "super"
Fixes #5358

peterberkenbosch committed:
set the order#payment_state to the correct state when order is canceled

Fixes #5278
Closes #5279
Closes #5388

return the new payment state in the update_payment_state, and add some spacing in the spec.

remove the stub to make sure this is not always returning "pending" *doh*

Senjai committed:
Update mail settings configuration change.

Prevents issues getting opened up all over the place about it.
Fixes #5367

AlessioRocco committed:
Add considered_safe to variable states

add translation for considered risky in order index

Add considered risky flag to orders index

Fixes #5374

camelmasa committed:
Bump Rails to 4.1.6

Fixes #5371

jordan-brough committed:
Revert "Added ability to perform address actions outside scope of an order"

This code is unreachable in stock spree because there is no route to get to Api::AddressesController without going through an order.
And the unreachable code is insecure by default -- if a store did add an "/api/address/:id" route then any Api user would be able to read any other user's addresses simply by guessing the id because that's how the permissions were added in ability.rb.
It seems like it would be better to just remove this or else make it a fully supported, reachable, and secure endpoint in stock Spree.

This reverts commit 229cdf4284fa56528a06c0c86830b19a7f49c3df.

Conflicts:
	api/app/controllers/spree/api/addresses_controller.rb
	core/app/models/spree/ability.rb

Fixes #5375

Commit:
OrderPopulator doesnt have to call shipment update as OrderContents handles that.

Commit:
Use email regex from Devise. Fixes #5368

HoyaBoya committed:
update order shipment total on transition to delivery

Fixes #5376

Commit:
Add data-hook to password fields on user form
Fixes #5377

Commit:
Fix spec broken by ca5799d2c1b0d3f5a2ea1b695e5c7083781080cd

reinaris committed:
Added failed state to the admin variables

A payment state could be `failed`.
This state was missing from the admin variables.
The payment state had no color, it was just a blank circle. I added the red color.
I noticed this in spree 2-2-stable.

Fixes #5378

metade committed:
Only present promotion codes that have been applied in the admin interface, and use pluck rather than collect to load the promotion name and id

Fixes #5379

Commit:
Don't process refunds created with a transaction_id

If a transaction id exists, presumably it has already been processed.
This comes in handy when migrating over records from another system.

Fixes #5381

Commit:
Fix spec broken by d0615f8

Commit:
Bump active merchant and fix broken spec.

BenMorganIO committed:
add select2 call after loading new images
Fixes #5385

andrzejpolis committed:
Regression spec for 'stack level too deep' #5384

Fixes #5384

Commit:
we fixed mail validator.

Commit:
Update readme for rails version.

reinaris committed:
Deleted the continue button from adjustment index

On the adjustment index (in the admin) was a continue button.
The button only shows up on this page, it's a leftover from the old 'manual add an order' proces in the admin.

I deleted it.

Fixes #5395

reinaris committed:
Close a table in variant autocomplete

There was no table closing tag in the variants autocomplete.
Fixes #5396

lipenco committed:
data-hook for form

Fixes #5397

reinaris committed:
Right no objects found message for new stock transfer
Fixes #5400

BenMorganIO committed:
add array & hash option for spree preferences

Fixes #5405
Fixes #5002

reinaris committed:
Deleted the ajax_indicator for the observe_field on change

Deleted the spinner from the states index
Fixes #5402

reinaris committed:
Added configuration_menu to the 'add country' page
Fixes #5403

reinaris committed:
Added configuration_menu to the 'stock locations edit'

Added configuration_menu to 'stock locations new'

Fixes #5404

reinaris committed:
Fixed objects name for stock location form labels
Fixes #5406

Commit:
Fix the damn taxes.  When tax adjustments are destroyed we need to set additional_tax_total & included_tax_total on items back to 0.

Fixes #5392

HoyaBoya committed:
Code for tax category.

Fixes #5409

hnatt committed:
Fix translation
Fixes #5408

hnatt committed:
Fix db:bootstrap task
Fixes #5431
Fixes #5432

BenMorganIO committed:
remove deprecated spree preferences methods from guide

introduce preferences column in spree models

add array and hash to spree preference types
Fixes #5410

frahugo committed:
Editing a shipping method was triggering to GET requests to the edit action

Add some missing preventDefault() and 'return false'
Fixes #5412

brchristian committed:
Standardizing use of colons in checkout summary
Fixes #5413

brchristian committed:
refactor shipping calculator compute methods

clearer naming for methods

Fixes #5414

Commit:
Dont let order import fail silently on bad line items.

marcosteixeira committed:
User store location when signing in
Fixes #5418
Fixes #5125

Conflicts:
	core/lib/spree/core/controller_helpers/auth.rb

Commit:
Fix my mistakes from 933c773a234048b72fbd7cb4fccce8810ec55c83 :(

Commit:
Remove unused check that stubs out valid data.

brchristian committed:
Include all adjustments in final shipment price

(Not solely promotions and taxes.) As per discussion in https://github.com/spree/spree/issues/5415

Fixes #5415
Fixes #5419

Commit:
Touch Adjustment#adjustable, and fix some rspec deprecations.

cgallegu committed:
Closes #5422

Commit:
Small performance improvement calculating taxes.

sanchojaf committed:
fix spec path in readme
Fixes #5423

BenMorganIO committed:
correct for value to match input id

Closes #5424

Commit:
refactoing code to call find_adjustment in before_action rather than calling it in each menthod
Fixes #5426

Commit:
fix regexp for credit card, also set month only if regexp is correct

Closes #5427

Commit:
Start release notes.

Commit:
Fix specs broken by 09ccf8a

Commit:
Fix specs broken by 87b3fa6

Commit:
Fix broken specs from 4441fb42acfbf7ebf02a44c116f0d09d61c3b54f & c3a741c

Commit:
Add spec to ensure adjustment callback is fired.

Commit:
Add comment explaining its important to use destroy_all.

Commit:
Replace scoped with where(nil) as scoped was removed from Rails. See: http://stackoverflow.com/a/18199294 Fixes #5366

mbj committed:
Fix factories to generate unique SKUs

* Choosing random SKUs even from a high range can yield duplicates
  resulting in build failures.
* Kernel.rand(9999) was definitively never a good choice due the
  birthday paradoxon lets manifest a duplicate quite often
* Uses sequencially generated SKUs defined from a single shared and
  reusable generator.

Fixes #5452

Commit:
Fix spec broken by 3eeb71d and add a regression check for it.

demetrodon committed:
fixed style quotes to path gem spree_core
fixes #5435

jordan-brough committed:
Remove some bad factory values

These numbers are autogenerated by the models so we don't need to set them in the factories.
On top of that, the numbers were being set 1) without their normal prefixes, 2) to the exact same value for every instance of the factory, 3) to the exact same values in two different factories.  These things caused errors in our app specs due to cross-model collisions that were a bit of a pain to track down.

Fixes #5437

HoyaBoya committed:
Allow order admin to filter on risky
Fixes #5438

Commit:
Some updates to release note links.

mbj committed:
Refactor Promotion.applied scope to query faster

* Before the scope did two joins. Joining the join table and than the
  orders table. As joining the join table is sufficient to define the
  result set removing one layer of joins significantly speeds up the
  operation on large order tables.

Fixes #5453

HoyaBoya committed:
added code to shipping methods

better layout

place shipping_method.code on selected shipping rate

Fixes #5439

magnusvk committed:
Associate new exchange orders with the same store as the original order

Replaces #110.

Fixes #5440

Commit:
Raise an error if no unprocessed payments.

Fixes #5436

/cc @athal7 This breaks an exchanges rake task spec spec.
I spent a bunch of time trying to make it work right,
but haven't figure out proper way to do it.  It appears the while order.next
loop completes order because of default credit card bug.  Basically need a solution
to: Refs #5327

Commit:
War on class_eval. refs #5358

Commit:
Mention we accept PRs for documentation.

Commit:
Add in fix from @andrzejpolis Fixes #5384 Fixes #5386

Commit:
Fix specs broken by e6865c6

Commit:
Fix guides syntax.

BenMorganIO committed:
label values for payment methods form

remove unused fullwidth class

It appears the button was loading as `<button class="fa fa-plus button"
name="button" type="{:class=>&quot;fullwidth&quot;}">Add</button>`
which isn’t very pretty.

bind numerous label tags to their inputs

add labels and move inputs onto single row

The quantity input was taking u pa whole row and so was the variant
input. A bit too long to look nice. Made it a lot more aesthetic and
add the label tags.

tab width with 2 spaces, not 4 spaced tab

use and instead of &amp;

the &amp; was not encoding, so I just switched to and.

add a page title for taxons

add taxons to products tab

update the adjustment button style

get more of the label tags matching

Fixes #5446

peterberkenbosch committed:
create proposed shipments if is no shippment data
also make sure to set the cost provided after order update

Conflicts:
	core/lib/spree/core/importer/order.rb
	core/spec/lib/spree/core/importer/order_spec.rb

mbj committed:
Fix example to contain expectation
Fixes #5449

pdamer committed:
allow guests to use the payments api for their orders
Fixes #5450

Commit:
Update basic RMA guide.

hnatt committed:
Improve localization of stock_locations/_form
Fixes #5454

StupidCodeFactory committed:
Update jquery-ui-rails

Fixes #5462

StupidCodeFactory committed:
Update dependencies

* updated kaminari, cancancan and Ransack
* fix product admin page:
  delete 'deleted_at_null' from the searchg query hash as now ransack behaves
  property and won't include products with null delete_at value

Fixes #5455

divineforest committed:
Fix problems with “database: true” on Teatro
Fixes #5457

HoyaBoya committed:
Make create endpoint require important params

bit of clean up on these tests, but no more after #create

Fixes #5458

Commit:
fix spelling for destroy at various places
Fixes #5459

Commit:
Remove extendable line items from notes.

Commit:
Fix regression from 9d12f0e

Commit:
Image configuration is no longer done.

Commit:
Stop running next! before complete step to fix broken spec.

metade committed:
Moving admin index route to bottom of routes file to allow it to be overridden without breaking pagination on the orders page

Fixes #5461

jordan-brough committed:
Refactor some ReimbursementTypeEngine code

No behavior change.
Extract a method to make it easier for subclasses to modify behavior by overriding just the "calculate_reimbursement_type" method.
Fixes #5465

HoyaBoya committed:
Add ability to track admin order canceler.

Fixes #5464

pdamer committed:
fix handling of credit cards expiring in current month
fixes #5468

ericsaupe committed:
Remove 'disabled' from inputs in payments

With the Aug 27 commit, a601035bd199a49f7f74ccf04ed0d889ca895f3e, the inputs under `.payment-method` are being disabled on line 15 but never correctly undisabled. This fixes that issue. Looks like it is affecting most branches.
fixes #5467

Commit:
Add Heroku deploy button.

Commit:
Update app.json

Commit:
Add buildpack url.

Commit:
Remove deploy to heroku button. Fixes #5401

Commit:
Add issuestats badges.

Commit:
fix spelling for interim in image model--skip-ci
fixes #5469

Commit:
Fix running just rake for specs.

Commit:
Fix rerunning sample specs without new dummy app.

Commit:
Assign store id to each order.

Commit:
Remove check thats no longer necessary.

Commit:
Bump version.

Commit:
Revert "Add Spree instrumentation"

This reverts commit 48669ed26f9ae7e1f07a31f19fbce794ce44ef9c.

Commit:
Bump version.

Commit:
Bump to rc3

Commit:
Check payment method should respond to cancel. Fixes #5307

Commit:
Improve the presentation during creating prototypes. Fixes #2526

jordan-brough committed:
Raise an exchange error if we can't ship all requested items

This will prevent an exchange from failing silently when, for example, a stale return authorization admin page gets submitted requesting an expedited exchange for an item for which there is no longer any stock available.

Fixes #5472

Commit:
fix spelling of retrieve everywhere
fixes #5473

BenMorganIO committed:
update readme for rails 4 asset precompilation
fixes #5474

BenMorganIO committed:
create a has and belongs to many association for spree taxons and prototypes

add taxons for admin prototypes

add spec to prototypes feature for deleting

remove old `.find_all_by_prototype` code

This code seems old. It is trying to do a table join on a non-existent
relationship, or rather, a table that is non-namespaces with `spree_`.
Also, I couldn’t get it working while writing tests for it. Tested out
the method on a current project and returned PG errors. Tests returned
SQLite errors.

Fixes #5476

BenMorganIO committed:
add carmen as a dependency

BenMorganIO committed:
create all the countries via carmen

BenMorganIO committed:
use carmen to create all the states

BenMorganIO committed:
implement active record base transaction to make it faster

BenMorganIO committed:
have states required return yes if country has subregions
fixes #5477
fixes #4435

BenMorganIO committed:
move tab arrays into backend configuration model via constants

This one is a really cool one because that means that if you’re using a
spree extension and you want to make a tab active, you don’t have to
overwrite the _tabs file. All you have to do is add to the constants
:D. This is awesome.

fixes #5478

Commit:
Default to use biling address on admin order form.

Commit:
Newer rspec syntax.

Commit:
Use valid zip codes to fix specs.

Commit:
No need for partial. Fixes #5475

ifan-godaddy committed:
Do not allow leading or trailing dot(.) in the first part of email address

Add rspec for email validator

Fixes #5483

Commit:
Remove shoulda matchers again.

alexstoick committed:
Looking up stock location by the admin name as well

fixes #5484

Commit:
Add spec for looking up stock location by admin name.

Commit:
Remove spec invalidated by https://github.com/spree/spree/commit/a13524a6cfd363c441ff1607a584a3be07e1124f now creating shipments.

Commit:
Messy merge.

Commit:
Provide @variant_images in products_controller

Commit:
Remove unused partials

Commit:
Image table improvements

- mark images for deletion
- show "no images" row if none are present

Commit:
Reveal controls for product form sections

Commit:
Add new image to table

Commit:
Merge branch '5010-redesign-product-form' of github.com:isaacfreeman/spree into 5010-redesign-product-form

Conflicts:
	backend/app/assets/javascripts/spree/backend/admin.js.erb
	backend/app/assets/stylesheets/spree/backend/sections/_products.scss
	backend/app/controllers/spree/admin/products_controller.rb
	backend/app/views/spree/admin/product_properties/index.html.erb
	backend/app/views/spree/admin/products/_form.html.erb
	backend/app/views/spree/admin/products/_images.html.erb
	backend/app/views/spree/admin/products/_variants.html.erb
	backend/app/views/spree/admin/products/_variants_table.html.erb
	backend/app/views/spree/admin/variants/_master_variant_fields.html.erb

Commit:
Fixes issue where form wasn't saving

Commit:
Moves stock management to main product page

Still a separate form at this stage.

Commit:
Fixes toggle for image fieldset

Commit:
Move some info and controls from stock items table to variants table

I eventually want to remove the stock items table entirely.

- edit track_inventory from variants table
- edit backorderable for stock_items from variants table
- variant accepts nested attributes for stock_items

Commit:
Improves variants table

- style adjustments for stock columns in variant table
- removes count_on_hand from legacy stock table
- new partial stock_item_show for stock item cells in variants table
- fixes some issues with edit row layout

Commit:
Product form layout and stock management improvements

- stock info moved into variant rows (add stock form still separate)
- manage stock for master variant
- toggle stock table when track inventory checkbox changes
- product edit fields resize vertically
- layout changes to add stock form
- seo fields moved alongside other product details

Commit:
Add stock per variant

- stock_item accepts nested attricutes for stock_movements
- styling for stock items table
- removes add stock form
- removes stock edit page
- modify quantity field for each stock item in table (creates a stock_movement if value is non-zero)
- moves weight and dimensions into sku column to make more horizontal room

Commit:
Manage images per variant

Commit:
Handlebar templates for image forms

Commit:
Tidying up images

- consolidates JS for new images

Commit:
Improved styling

Commit:
Improved stock management

- reordered admin.js
- added styles for variant images
- new partial images_show
- empty state for tables, with add link
- changed option columns in variants table to single column to conserve horizontal space
- changed layout for stock item partial
- fixes to updating show fields when edit fields change
- variants only create stock_items if they're not already there.

Commit:
Manage product properties

- add/edit/remove/reorder product properties with Handlebars template
- update display of stock information when track_inventory checkbox is changed
- fixes reordering images
- display error messages on product's nested objects
- reject images with no attachments
